Include mod effects in values displayed by `BeatmapAttributeText`

C0D3-M4513R opened the pull request:
closes https://github.com/ppy/osu/issues/21946

Applying a mod with a difficulty change did previously not update the values displayed in BeatmapAttributeText. This attempts to work around this, by taking heavy "inspiration" from `osu.Game.Screens.Select.Details.AdvancedStats` and `osu.Game.Screens.Select.BeatmapInfoWedge`.

Now the displayed mod statistics are the same in the Beatmap details section in the select screen, and in the values in BeatmapAttributeText.

One thing to keep in mind is that Length will currently not get updated at all, but that seems to be a known restriction currently.

C0D3-M4513R committed:
Applying mods did not update values in BeatmapAttributeText

Applying a mod with a difficulty change did previously not update the values displayed in BeatmapAttributeText.
This attempts to work around this, by taking heavy "inspiration" from `osu.Game.Screens.Select.Details.AdvancedStats` and `osu.Game.Screens.Select.BeatmapInfoWedge`.

Now the displayed mod statistics are the same in the Beatmap details section in the select screen, and in the values in BeatmapAttributeText.

C0D3-M4513R committed:
Also apply fixed rate adjustment to Beatmap Length

bdach commented on osu.Game/Skinning/Components/BeatmapAttributeText.cs:
```diff
@@ -34,8 +38,27 @@ public partial class BeatmapAttributeText : Container, ISkinnableDrawable
         [SettingSource("Template", "Supports {Label} and {Value}, but also including arbitrary attributes like {StarRating} (see attribute list for supported values).")]
         public Bindable<string> Template { get; set; } = new Bindable<string>("{Label}: {Value}");
 
+        [SettingSource("Adjust Difficulty", "Should Mods that change Beatmap Statistics be applied to the shown Stats?")]
+        public BindableBool DifficultyAdjust { get; set; } = new BindableBool(true);
```
Should not be a setting. Either this PR is the wrong direction entirely and none of this is necessary, or the component as it is in `master` is plain wrong and this diff is required. Having this toggleable helps nobody.

bdach replied:
This was briefly [discussed in discord](https://discord.com/channels/188630481301012481/188630652340404224/1059872767555682405) today, and either this setting should not exist (and all values should be post-mod application), or there should be 2 sets of variables offered for pre- and post-mod values.

bdach replied:
Well yes. The current implementation would probably have to somewhat improve as it is currently mighty inefficient and would only get worse. But it covers more use cases than the switch.

bdach commented on osu.Game/Skinning/Components/BeatmapAttributeText.cs:
```diff
@@ -34,8 +38,27 @@ public partial class BeatmapAttributeText : Container, ISkinnableDrawable
+        [Resolved]
+        private IBindable<IReadOnlyList<Mod>> mods { get; set; } = null!;
+
+        private ModSettingChangeTracker modSettingChangeTracker = null!;
```
Should be annotated as nullable (as it initially is `null`, after all).

bdach commented on osu.Game/Skinning/Components/BeatmapAttributeText.cs:
```diff
@@ -78,27 +103,106 @@ protected override void LoadComplete()
+            starDifficultyTask.ContinueWith(_ => Schedule(() =>
+            {
+                starRating = starDifficultyTask.GetResultSafely();
+                valueDictionary[BeatmapAttribute.StarRating] = (starRating?.Stars ?? -1).ToLocalisableString(@"F2");
```
Why is -1 used as the fallback value? This will be printed to the user. May as well use something like `"(not available)"` or anything.

bdach commented:
I will also say yet again that this component is getting very huge and complicated *very* quickly. Not happy to see increasing scope creep in this area from multiple directions.

C0D3-M4513R committed:
Make modSettingChangeTracker nullable

C0D3-M4513R committed:
Create new `GetBPMAndLength` in `BeatmapExtensions`

This simplifies parts of `BeatmapInfoWedge.refreshBPMLabel` and `BeatmapAttributeText.updateBpmAndLength`.

C0D3-M4513R committed:
Change default SR in `BeatmapAttributeText.updateStarRating`

This changes how the StarRating is Set in `BeatmapAttributeText`. Previously the default StarRating, before any Calculation had taken place was `-1`, because of `BeatmapInfo.StarRating`.
Now we use that StarRating implicitly through the already set Beatmap.

C0D3-M4513R committed:
Remove accidental extra `0` in `BeatmapAttributeText`

C0D3-M4513R committed:
Fix BPM calculation not staying the same

mk56-spn commented:
@C0D3-M4513R 

Sorry for the wait, here ya go.

https://user-images.githubusercontent.com/74463310/207155164-ac13c855-3aee-420c-a516-2a9e14315a9d.mp4

mk56-spn commented:
> Can you please retest with changes in this pr added? I see from the settings that you are using regular lazer, which still has this issue.

all good now, sorry for the troubles

C0D3-M4513R committed:
Remove DifficultyAdjust, and add more enum Values instead

Also fixes some of the bpm values being wrong.

C0D3-M4513R committed:
Remove useless Comment

C0D3-M4513R committed:
Merge branch 'master' into beatmap-info

C0D3-M4513R committed:
Remove `#nullable disable` in StarDifficulty

C0D3-M4513R committed:
Remove `#nullable disable` in StarDifficulty

C0D3-M4513R committed:
Only compute values as needed.

C0D3-M4513R committed:
Only compute values as needed.

C0D3-M4513R committed:
Merge remote-tracking branch 'origin/beatmap-info' into beatmap-info

# Conflicts:
#	osu.Game/Skinning/Components/BeatmapAttributeText.cs

C0D3-M4513R committed:
Apply CodeFactor suggestions

C0D3-M4513R committed:
Merge branch 'master' into beatmap-info

bdach commented:
> I wanted to make a new global cache that just saves these things and reuses them for `BeatmapAttributeText` and `BeatmapInfoWedge`

I don't really have the mental energy to argue over this too much - especially so that I was against merging this feature from the start precisely due to the scope creep that it has provoked - but I still don't think anything global like that is necessary.

I consider this PR as-is unmergeable, as it is adding a lot of complication to everything because `BeatmapAttributeText` is trying to coordinate everything instead of modularising the value fetch logic appropriately (which, again, *doesn't require global caching*). I described [a sketch](https://discord.com/channels/188630481301012481/188630652340404224/1062444624129560706) of how I imagine this working on discord but if I were to go into any more detail than that I'd end up writing the entirety of the solution which I don't want to do right now.

C0D3-M4513R committed:
Use `Lazy<Task<LocalisableString>>`
This allows us to ensure that all computation happens only when needed, and to async some calculations.

C0D3-M4513R committed:
Fix StarRating not Updating correctly.

C0D3-M4513R committed:
Only computes Values once.

This makes `values` a static member, and uses tricks to only compute the entries in that `values` dictionary once.
Currently `values` doesn't get disposed when not needed anymore. Comments about that have been made in the `Dispose` method.

Signed-off-by: C0D3 M4513R <28912031+C0D3-M4513R@users.noreply.github.com>

C0D3-M4513R committed:
Allow for `values` to be deallocated, when needed

C0D3-M4513R committed:
Switch to a better approach of knowing when the last BeatmapAttributeText get's disposed.

C0D3-M4513R committed:
Remove default value.

I am beginning to sometimes not like nvika.

bdach commented:
There are too many contributions for me to get a thorough review on this as I consider this a low priority task. However at a very quick look at the current state I already don't think I'm going to like what is here now. The nine-something added members and +300 lines in `BeatmapAttributeText` alone tell me that this is likely not structured and modularised as I'd like to see it, and I cannot provide any better guidance without going to write this myself,  which I do not want to do at this time.

peppy commented:
This isn't the way forward. There's just too much complexity, as has already been stated. Needs to be reapproached without changing the whole structure of the component. It should not be required.
